src/type-coercions.md

```diff
@@ -182,5 +180,109 @@ unsized coercion to `Foo<U>`.
 > has been stabilized, the traits themselves are not yet stable and therefore
 > can't be used directly in stable Rust.
 
-[`Unsize`]: ../std/marker/trait.Unsize.html
+## Least upper bound coercions
+
+*Least upper bound coercion* (LUB coercion) is the form of type coercion take
+place when several expressions of possibly different types need be unified to a
+single type. This happens in [match] arms, [if expressions] and [array literal
+expressions]. In LUB coercion, the compiler tries to find the least upper
+bound of given types.
+
+For example:
+
+```rust
+# let (a, b, c) = (0, 1, 2);
+let foo = if true {
+    a
+} else if false {
+    b
+} else {
+    c
+};
+
+let bar = match 42 {
+    0 => a,
+    1 => b,
+    _ => c,
+};
+
+let baz = [a, b, c];
+```
+
+In this example, both `foo` and `bar` have the type
+`LubCoerce(typeof(a), typeof(a), typeof(c))` and `baz` have the type
+`[LubCoerce(typeof(a), typeof(b), typeof(c)); 3]`.
+
+LUB coercion is performed by the following algorithm:
+
+```txt
+Lub(a: Type, b: Type):
+  if a == (FnDef | Closure)
+    && b == (FnDef | Closure)
+    && a != capturing Closure
+    && b != capturing Closure:
+    return FnPtr
+
+  // Coerce(x, y) returns true when x can be coerced to y by coercion described
+  // in previous sections of this page.
+  if Coerce(b, a):
+    return a
+  if Coerce(a, b):
+    return b
+
+  // LubCoerce failed
+  emit type error
+
+LubCoerce(vars...):
+  result = vars.get(0)
+  for var in vars[1..]:
+    result = Lub(result, var)
+  return result
+```
+
+LUB coercion has the following properties:
+
+1. Order independent: e.g. `LubCoerce(ty0, ty1, ty2, ty3)` equals to
+    `LubCoerce(ty1, ty0, ty4, ty3)`.
+2. `LubCoerce(ty0, ty1, ty2, ...)` equals to
+    `LubCoerce(LubCoerce(ty0, ty1), ty2, ...)`
+3. `LubCoerce(ty0, ty1) == Some(ty2)` means both `ty0` and `ty1` can be coerced
+    to `ty2`.
+
+Note the feature No. 3, it uses the word "means" rather than "if and only if".
+That's because currently if `ty0` and `ty1` can be coerced to `ty2` and
+unfortunately `ty2` equals to neither `ty0` nor `ty1`, there are only one
+special situation where we can get `LubCoerce(ty0, ty1) == Some(ty2)`:
+`LubCoerce((FnDef | Closure), (FnDef | Closure)) == Some(FnPtr)` (where Closure
+is non-capturing). You can check it with the pseudo code.
```

Comment:
I don't really know what this sentence is trying to say.

```suggestion
is non-capturing).
```

Reply:
Thanks! This sentence is not necessary though.